src/number_theory/bernoulli_polynomials.lean

```diff
@@ -0,0 +1,163 @@
+/-
+Copyright (c) 2021 Ashvni Narayanan. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Ashvni Narayanan
+-/
+
+import number_theory.bernoulli
+
+/-!
+# Bernoulli polynomials
+
+The Bernoulli polynomials are an important tool obtained from Bernoulli numbers.
+
+## Mathematical overview
+
+The $n^ℐrm{th}$ Bernoulli polynomial is defined as
+$$ B_n(X) = ∑_{k = 0}^n {n χoose k} (-1)^k * B_k * X^{n - k} $$
+where $B_k$ is the $k^ℐrm{th}$ Bernoulli number. The Bernoulli polynomials are generating functions,
+$$ t * e^{tX} / (e^t - 1) = ∑_{n = 0}^{∞} B_n(X) * ¼{t^n}{n!} $$
+
+## Implementation detail
+
+The Bernoulli (negative) numbers are first defined, `bernoulli_neg`, in order to make proving
+theorems using Bernoulli polynomials easier.
+
+## Main theorems
+
+- `sum_bernoulli_poly`: The sum of the $k^\mathrm{th}$ Bernoulli polynomial with binomial
+  coefficients up to n is `(n + 1) * X^n`.
+- `exp_bernoulli_poly`: The Bernoulli polynomials act as generating functions for the exponential.
+-/
+
+noncomputable theory
+open_locale big_operators
+open_locale nat
+
+open nat finset
+
+/-- The Bernoulli polynomials are defined in terms of the negative Bernoulli numbers. -/
+def bernoulli_poly (n : ℕ) : polynomial ℚ :=
+  ∑ i in range (n + 1), polynomial.monomial (n - i) ((bernoulli i) * (choose n i))
+
+lemma bernoulli_poly_def (n : ℕ) : bernoulli_poly n =
+  ∑ i in range (n + 1), polynomial.monomial i ((bernoulli (n - i)) * (choose n i)) :=
+begin
+  rw [←sum_range_reflect, add_succ_sub_one, add_zero, bernoulli_poly],
+  apply sum_congr, {refl,},
+  rintros x hx,
+  rw mem_range_succ_iff at hx, rw [choose_symm hx, nat.sub_sub_self hx],
+end
+
+namespace bernoulli_poly
+
+/-
+### examples
+-/
+
+section examples
+
+@[simp] lemma bernoulli_poly_zero : bernoulli_poly 0 = 1 :=
+by { rw bernoulli_poly, simp, }
+
+@[simp] lemma bernoulli_poly_zero' (n : ℕ) : (bernoulli_poly n).eval 0 = bernoulli n :=
+begin
+  rw [bernoulli_poly, polynomial.eval_finset_sum], simp only [polynomial.eval_monomial],
+  rw sum_range_succ,
+  simp only [add_right_eq_self, mul_one, cast_one, nat.sub_self, choose_self, pow_zero],
+  apply sum_eq_zero, rintros x hx,
+  rw [zero_pow', mul_zero], apply mem_range_sub_ne_zero hx,
+end
+
+end examples
+
+@[simp] theorem sum_bernoulli_poly (n : ℕ) :
+  ∑ k in range (n + 1), ((n + 1).choose k : ℚ) • bernoulli_poly k =
+    polynomial.monomial n (n + 1 : ℚ) :=
+begin
+  cases n,
```

Comment:
This is not necessary!

If you were doing induction and using the induction hypothesis somehow, a case split into `zero` and `succ` makes sense. But n=0 is not a special case here. If you delete it you can then tidy up your proof a bit because there aren't these irrelevant `succ`s everywhere. Here is a working version with no `cases` -- it's your proof of the `succ` case, and the observation that it works just as well for the 0 case (plus the corresponding simplification in your proof of `f` -- which is a funny name for a hypothesis ;-) ). I also changed the last `simp` because the result you want is in the library -- it's just `sum_const_zero`. 

```
  simp_rw [bernoulli_poly_def, finset.smul_sum, finset.range_eq_Ico, ←finset.sum_Ico_Ico_comm,
    finset.sum_Ico_eq_sum_range],
  simp only [cast_succ, nat.add_sub_cancel_left, nat.sub_zero, zero_add, linear_map.map_add],
  simp_rw [polynomial.smul_monomial, mul_comm (bernoulli _) _, smul_eq_mul, ←mul_assoc],
  conv_lhs { apply_congr, skip, conv
    { apply_congr, skip,
      rw [nat.prime.choose_mul ((nat.le_sub_left_iff_add_le (mem_range_le H)).1 (mem_range_le H_1))
        (le.intro rfl), add_comm x x_1, nat.add_sub_cancel, mul_assoc, mul_comm, ←smul_eq_mul,
        ←polynomial.smul_monomial], },
    rw [←sum_smul], },
  rw sum_range_succ,
  simp only [add_right_eq_self, cast_succ, mul_one, cast_one, cast_add, nat.add_sub_cancel_left,
    choose_succ_self_right, one_smul, bernoulli_zero, sum_singleton, zero_add,
    linear_map.map_add, range_one],
  have f : ∀ x ∈ range n, 2 ≤ n + 1 - x,
  { rintros x H, rw [mem_range] at H,
    exact nat.le_sub_left_of_add_le (succ_le_succ H) },
  exact sum_const_zero,
```

Reply:
Noted and changed, thank you!